src/org/traccar/protocol/NavisFrameDecoder.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019 Anton Tananaev (anton@traccar.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.TooLongFrameException;
+import java.nio.charset.StandardCharsets;
+import org.traccar.BaseFrameDecoder;
+
+public class NavisFrameDecoder extends BaseFrameDecoder {
+
+    private static final int NTCB_HEADER_LENGTH = 16;
+    private static final int NTCB_LENGTH_OFFSET = 12;
+    private static final int FLEX_HEADER_LENGTH = 2;
+    private static final int MAX_FRAME_LENGTH = 65551;
+
+    private final NavisProtocolDecoder protocolDecoder;
+
+    public NavisFrameDecoder(NavisProtocolDecoder protocolDecoder) {
```

Comment:
Just noticed that you pass protocol decoder into frame decoder. That's not how we usually implement things. You need to get decoder from pipeline.

Reply:
Similar code I use also in frame decoder tests where there is no pipeline. Please point me to the example in the current codebase so that I can use the case with pipeline.

Reply:
For unit tests we usually use setters. So, if you need to modify frame decoder state - use setters.

Reply:
I suggest you have an instance variable for `FlexDataSize`. If we have pipeline, take the value from there originally. For unit tests you just use setter to set it.

Reply:
I suppose what you had in mind is now realized.